File: 0947-most-stones-removed-with-same-row-or-column/0947-most-stones-removed-with-same-row-or-column.js
/**
 * @param {number[][]} stones
 * @return {number}
 */
var removeStones = function(stones) {
  const parent = {};
  let count = 0;
  
  const find = (x) => {
    if (!parent[x]) {
      parent[x] = x;
      count++;
    }
    
    else if (x !== parent[x]) parent[x] = find(parent[x]);
    
    return parent[x];
  }
  
  const union = (x, y) => {
    const rootX = find(x);
    const rootY = find(y);
    if (rootX === rootY) return;
    
    parent[rootX] = rootY;
    count--;
  }
  
  stones.forEach(([u, v]) => union(u, ~v));
  
  return stones.length - count;
};

File: 0947-most-stones-removed-with-same-row-or-column/0947-most-stones-removed-with-same-row-or-column.py
class Solution:
    def removeStones(self, stones: List[List[int]]) -> int:
        parent = {}
        def find(x):
            if not x in parent:
                parent[x] = x
            elif x != parent[x]:
                parent[x] = find(parent[x])
            return parent[x]
        
        
        for u, v in stones:
            parent[find(u)] = find(~v)
        return len(stones) - len({find(x) for x in parent})
            